### PR TITLE
Fixed #32227 -- Prevented crash when setUpTestData() errors with --debug-sql.

### PR DESCRIPTION
**Pull request by adamchainz**
ticket-32227

Thanks Mariusz Felisiak for the report.

I don't have PostgreSQL setup for django testing locally so I reproduced instead with:

```patch
diff --git a/tests/timezones/tests.py b/tests/timezones/tests.py
index 8756f87716..3924da4e1a 100644
--- a/tests/timezones/tests.py
+++ b/tests/timezones/tests.py
@@ -1162,6 +1162,8 @@ class AdminTests(TestCase):
             email='super@example.com', is_staff=True, is_active=True,
             date_joined=datetime.datetime(2007, 5, 30, 13, 20, 10, tzinfo=UTC),
         )
+        with connection.cursor() as cursor:
+            cursor.execute("SELECT SELECT")
 
     def setUp(self):
         self.client.force_login(self.u1)
```

### CODE REVIEW

**Comment by felixxm** [comment]
@adamchainz Thanks :+1: Is this testable?

**Comment by adamchainz** [comment]
I was just figuring that out... yes, kind of. I've modified the tests so there's at least coverage, but not sure the assertion on the output line is the best.

**Comment by carltongibson** [comment]
Speedy work. Thank you. (Carry on 🙂)

File: tests/test_runner/test_debug_sql.py [hunk 1]
```diff
@@ -90,12 +98,13 @@ def test_output_verbose(self):
...
             '''FROM "test_runner_person" WHERE '''
             '''"test_runner_person"."first_name" = 'subtest-fail';'''),
+        'ERROR: setUpClass (test_runner.test_debug_sql.TestDebugSQL.ErrorClassTest)',
     ]
 
     verbose_expected_outputs = [
+        'runTest (test_runner.test_debug_sql.TestDebugSQL.PassingTest) ... ok',
         'runTest (test_runner.test_debug_sql.TestDebugSQL.FailingTest) ... FAIL',
         'runTest (test_runner.test_debug_sql.TestDebugSQL.ErrorTest) ... ERROR',
-        'runTest (test_runner.test_debug_sql.TestDebugSQL.PassingTest) ... ok',
```
Comment (felixxm) [inline tests/test_runner/test_debug_sql.py:98]: This looks unrelated.

File: tests/test_runner/test_debug_sql.py [hunk 2]
```diff
@@ -25,6 +25,13 @@ def runTest(self):
             Person.objects.filter(first_name='error').count()
             raise Exception
 
+    class ErrorClassTest(TestCase):
+        def setUpTestData(cls):
```
Comment (felixxm) [inline tests/test_runner/test_debug_sql.py:29]: `@classmethod`

File: tests/test_runner/test_debug_sql.py (hunk 1 above)
Comment (adamchainz) [reply tests/test_runner/test_debug_sql.py:98]: Yes fair. I was trying to rearrange to match the orders but now I see `PassingTest()` is used in a different order to its definition.

File: tests/test_runner/test_debug_sql.py [hunk 3]
```diff
@@ -90,6 +99,7 @@ def test_output_verbose(self):
         ('''SELECT COUNT(*) AS "__count" '''
             '''FROM "test_runner_person" WHERE '''
             '''"test_runner_person"."first_name" = 'subtest-fail';'''),
+        'ERROR: setUpClass (test_runner.test_debug_sql.TestDebugSQL.ErrorClassTest)',
```
Comment (felixxm) [inline tests/test_runner/test_debug_sql.py:102]: This patch fixes the issue for me, however new tests work even without it, so I don't think we need them.

Unfortunately, it's more complicated then raising and error in the `setUpTestData()`.

Comment (felixxm) [reply tests/test_runner/test_debug_sql.py:102]: OK I figure out why it doesn't fail. `ErrorClassTest` must be the first one, because there is a single instance of `DebugSQLTextTestResult` so `debug_sql_stream` already exists  if we run any test before it. Edits is progress ...

Comment (felixxm) [reply tests/test_runner/test_debug_sql.py]: Crash depends on the order of added tests, so I moved it to a separate one.

Comment (adamchainz) [reply tests/test_runner/test_debug_sql.py]: Ah yes, of course! Thanks for the new test.

**Comment by adamchainz** [comment]
Thanks for fixing up the test and merging! 👍

**Comment by felixxm** [comment]
@adamchainz `test_runner.test_debug_sql.TestDebugSQL.test_setupclass_exception` fails when run in reverse:
```
./runtests.py test_runner --parallel=1 --reverse

Traceback (most recent call last):
  File "/home/jenkins/workspace/master-reverse/database/sqlite3/label/bionic/python/python3.9/tests/test_runner/test_debug_sql.py", line 132, in test_setupclass_exception
    self.assertIn(
AssertionError: 'ERROR: setUpClass (test_runner.test_debug_sql.TestDebugSQL.ErrorSetUpTestDataTest)' not found in '======================================================================\nERROR: runTest (test_runner.test_debug_sql.TestDebugSQL.ErrorSetUpTestDataTest)\n----------------------------------------------------------------------\nTraceback (most recent call last):\n  File "/home/jenkins/workspace/master-reverse/database/sqlite3/label/bionic/python/python3.9/django/test/testcases.py", line 272, in _setup_and_call\n    self._pre_setup()\n  File "/home/jenkins/workspace/master-reverse/database/sqlite3/label/bionic/python/python3.9/django/test/testcases.py", line 935, in _pre_setup\n    self._fixture_setup()\n  File "/home/jenkins/workspace/master-reverse/database/sqlite3/label/bionic/python/python3.9/django/test/testcases.py", line 1235, in _fixture_setup\n    self.setUpTestData()\n  File "/home/jenkins/workspace/master-reverse/database/sqlite3/label/bionic/python/python3.9/tests/test_runner/test_debug_sql.py", line 31, in setUpTestData\n    raise Exception\nException\n\n----------------------------------------------------------------------\n\n----------------------------------------------------------------------\nRan 0 tests in 0.000s\n\nFAILED (errors=1)\n'
```

The output is different:

- without `reverse`: `ERROR: setUpClass (test_runner.test_debug_sql.TestDebugSQL.ErrorSetUpTestDataTest)`
- with `reverse`: `ERROR: runTest (test_runner.test_debug_sql.TestDebugSQL.ErrorSetUpTestDataTest)`

Can you take a look?

**Comment by felixxm** [comment]
> Can you take a look?

I've prepared a patch, #13758.
